Extends email token documentation in emails.rst (#194)
mautic/developer-documentation-new: docs/components/emails.rst

There are multiple ways to extend the way Mautic works with Emails. This document describes the following options for extending Mautic's Email capabilities:

* Email tokens
* A/B testing
* Monitored Inbox Integration
* Email transport/Email providers
* Email stat helpers

.. vale off

.. note::
    
   Extending generally works by hooking into events using event listeners or subscribers. Read more about :doc:`listeners and subscribers</plugins/event_listeners>`.

.. vale on

Email tokens
************

> Email tokens are placeholders that you can insert into an Email. They get replaced by Dynamic Content once the Email gets sent or viewed in the browser.

Reviewer: ```suggestion
Email tokens are placeholders that you can insert into an Email. They get replaced by dynamically generated content once the Email gets sent or viewed in the browser.
```
Users shouldn't associate it with Dynamic Content. Although Dynamic Content is also using tokens it is just a subset of tokens.

Email token capabilities consist of two parts:

* Registering custom tokens
* Rendering custom tokens

Registering custom tokens in builders
=====================================

Registering tokens leverages the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_BUILD`` event. The event is dispatched before displaying the email builder form, to allow adding of tokens.

An event listener receives the ``Mautic\EmailBundle\Event\EmailBuilderEvent``.
Use its ``$event->addToken($token, $htmlContent)`` to add your token.

.. note::

   You can either hard-code your tokens' textual description in ``$htmlContent`` or use a translatable string.

Rendering custom tokens
=======================

> Registering tokens leverages the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_SEND`` event when the Email is sent or the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_DISPLAY`` event when the Email is rendered for viewing in a browser, such as after the Lead clicked the ``{webview_url}`` link.

Reviewer 2: This section is about **rendering** custom tokens, and not **registering** tokens.
I'm not quite sure about the technical here, but would below suggestion explain it well? Also, what does "Lead" refer to here?

```suggestion
To render custom tokens, use the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_SEND`` event when Mautic sends the Email, or the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_DISPLAY`` event when the Email displays in a browser such as after the Lead clicks the ``{webview_url}`` link.
```

Reviewer 3: A lead is a Contact - it's the old way of referring to them, so this should be updated to Contact.

I think the text was correct. The first part is talking about registering the token when the email is sent (e.g. this token exists, and it's this value) and then rendering the in the browser once the user clicks to open it. I might be misunderstanding, perhaps @mautic/core-team can review and provide input here.

Reviewer: Yes, @adiati98 is correct. The tokens for registering tokens are `EmailEvents::EMAIL_ON_BUILD` and `PageEvents::PAGE_ON_BUILD`.

Reviewer 2: I've applied the suggested change above and changed "Lead" to "Contact".

An event listener receives in both cases the ``Mautic\EmailBundle\Event\EmailSendEvent``.
You can replace a custom token using the events ``$event->addToken($token, $contentToReplaceToken)``.

Basic token replacement
=======================

.. code-block:: PHP

    <?php
    
    // plugins/HelloWorldBundle/EventListener/EmailSubscriber.php
    class EmailSubscriber implements EventSubscriberInterface
    {
    
      public static function getSubscribedEvents(): array
      {
        return [
          EmailEvents::EMAIL_ON_BUILD => ['onEmailBuild', 0],
          EmailEvents::EMAIL_ON_SEND => ['onEmailGenerate', 0],
          EmailEvents::EMAIL_ON_DISPLAY => ['onEmailGenerate', 0],
        ];
      }
    
      public function onEmailBuild(EmailBuilderEvent $event): void
      {
        $event->addToken('{my_custom_token}', 'My Custom Token');
      }
    
      public function onEmailGenerate(EmailSendEvent $event): void
      {
        $event->addToken('{my_custom_token}', 'Hello <b>World!</b>');
      }
    }

.. note::

>    If you need more complex replacing, you can use the event's ``$event->getContent()`` and ``$event->setContent()`` methods. See the example in the following section.

Reviewer 2: Would this be clearer?

```suggestion
   For more complex replacements, use the event's ``$event->getContent()`` and ``$event->setContent()`` methods.
```

.. vale off

Email A/B testing
*****************

.. vale on

While Mautic supports :xref:`A/B testing` out of the box, you might have more complex needs to determine A/B test winner criteria.

* ``$event->addAbTestWinnerCriteria()`` allows you to do exactly that. Using your custom logic, you can decide the winner of such criteria.

> * ``$event->setAbTestResults()`` is where you set the actual A/B test results.

Reviewer 2: Would the suggestions below still hold the same meaning?
 
```suggestion
* Use ``$event->addAbTestWinnerCriteria()`` to apply your custom logic when deciding the winner based on specific criteria.
* Set the actual A/B test results with ``$event->setAbTestResults()``.
```

Reviewer 3: I think so, yes.

- Email transport/Email providers
- Email stat helpers

.. Note:: Extending generally works by hooking into events using event listeners or subscribers. Read more about :doc:`listeners and subscribers</plugins/event_listeners>`.

Email tokens
------------

Email tokens are placeholders that you can insert into an Email.
They get replaced by Dynamic Content once the Email gets sent or viewed in the browser.

Email token capabilities consist of two parts: registering your custom tokens and rendering them.

Registering custom tokens in builders

> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Reviewer 3: We use the following:

H1 ===
H2 ***
H3 ---
H4 ---
H5 ~~~
H6 ^^^

I think these should therefore be using ~~~?

Author: Changed

> Registering tokens leverage the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_BUILD`` event.

Reviewer 3: ```suggestion
Registering tokens leverages the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_BUILD`` event.
```

> The event is dispatched before displaying the email builder form to allow adding of tokens.

Reviewer 3: ```suggestion
The event is dispatched before displaying the email builder form, to allow adding of tokens.
```

An event listener receives the ``Mautic\EmailBundle\Event\EmailBuilderEvent``.
Use its ``$event->addToken($token, $htmlContent)`` to add your token.

> .. Note:: You can either hard code your tokens textual description in ``$htmlContent`` or use a translatable string.

Reviewer 3: Please use the correct syntax here.

https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#notes

Rendering custom tokens
^^^^^^^^^^^^^^^^^^^^^^^

Registering tokens leverage the
- ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_SEND`` event when the email is sent or the

> - ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_SEND`` event when the email is rendered for viewing in a browser, i.e., after the Lead clicked the ``{webview_url}`` link.

Reviewer 3: Is this meant to be identical to the one above? Seems like it should be something different?

An event listener receives in both cases the ``Mautic\EmailBundle\Event\EmailSendEvent``.

> Replacing a custom token is as easy as using the events ``$event->addToken($token, $contentToReplaceToken)``.

Reviewer 3: ```suggestion
You can replace a custom token using the events ``$event->addToken($token, $contentToReplaceToken)``.
```

Let's not assume that just because we think something is easy, everyone else will: https://developers.google.com/style/tone#some-things-to-avoid-where-possible

> Example

Reviewer 3: Headings need to be unique, so please consider using something less generic here.

^^^^^^^

.. code-block:: PHP

    <?php
    
    // plugins/HelloWorldBundle/EventListener/EmailSubscriber.php
    class EmailSubscriber implements EventSubscriberInterface
    {
    
      public static function getSubscribedEvents(): array
      {
        return [
          EmailEvents::EMAIL_ON_BUILD => ['onEmailBuild', 0],
          EmailEvents::EMAIL_ON_SEND => ['onEmailGenerate', 0],
          EmailEvents::EMAIL_ON_DISPLAY => ['onEmailGenerate', 0],
        ];
      }
    
      public function onEmailBuild(EmailBuilderEvent $event): void
      {
        $event->addToken('{my_custom_token}', 'My Custom Token');
      }
    
      public function onEmailGenerate(EmailSendEvent $event): void
      {
        $event->addToken('{my_custom_token}', 'Hello <b>World!</b>');
      }
    }

> .. Note:: If you need more complex replacing you can use the event's ``$event->getContent()`` and ``$event->setContent()`` methods. See the example in the following section.

Reviewer 3: Please use the correct syntax (line break/indentation) for the note codeblock: https://sublime-and-sphinx-guide.readthedocs.io/en/latest/notes_warnings.html#notes

Email A/B testing
-----------------

While Mautic supports :xref:`A/B testing` out of the box, you might have more complex needs to determine A/B test winner criteria.

- ``$event->addAbTestWinnerCriteria()`` allows you to do exactly that. Using your custom logic, you can decide the winner of such criteria. An example is shown below.
- ``$event->setAbTestResults()`` is where you set the actual A/B test results. More details are in the code example below.

> Example

Reviewer 3: Headings need to be unique so please consider using something that's less generic here.

- Email transport/Email providers
- Email stat helpers

.. Note:: Extending generally works by hooking into events using event listeners or subscribers. Read more about :doc:`listeners and subscribers</plugins/event_listeners>`.

Email tokens

> ------------

Reviewer 3: ```suggestion
************
```

Please use the correct headings:

```
H1 ###
H2 ***
H3 ===
H4 ---
H5 ~~~
```

This is a H2 so it should be *** not ---

Email tokens are placeholders that you can insert into an Email.
They get replaced by Dynamic Content once the Email gets sent or viewed in the browser.

Email token capabilities consist of two parts: registering your custom tokens and rendering them.

Registering custom tokens in builders

> ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 3: See earlier comment about headings. This should be a H3 if you want it to be nested under the H2 above.

Registering tokens leverages the ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_BUILD`` event.
The event is dispatched before displaying the email builder form, to allow adding of tokens.

An event listener receives the ``Mautic\EmailBundle\Event\EmailBuilderEvent``.
Use its ``$event->addToken($token, $htmlContent)`` to add your token.

.. Note::
    You can either hard code your tokens textual description in ``$htmlContent`` or use a translatable string.

Rendering custom tokens

> ~~~~~~~~~~~~~~~~~~~~~~~

Reviewer 3: See above about headings.

Registering tokens leverage the
- ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_SEND`` event when the email is sent or the
- ``\Mautic\EmailBundle\EmailEvents::EMAIL_ON_DISPLAY`` event when the email is rendered for viewing in a browser, i.e., after the Lead clicked the ``{webview_url}`` link.

An event listener receives in both cases the ``Mautic\EmailBundle\Event\EmailSendEvent``.
You can replace a custom token using the events ``$event->addToken($token, $contentToReplaceToken)``.

> Basic Token Replacement

Reviewer 3: ```suggestion
Basic token replacement
```

Please don't use camel case - we use sentence case in headings.

> ^^^^^^^

Reviewer 3: Heading also needs fixing here please.

.. code-block:: PHP

    <?php
    
    // plugins/HelloWorldBundle/EventListener/EmailSubscriber.php
    class EmailSubscriber implements EventSubscriberInterface
    {
    
      public static function getSubscribedEvents(): array
      {
        return [
          EmailEvents::EMAIL_ON_BUILD => ['onEmailBuild', 0],
          EmailEvents::EMAIL_ON_SEND => ['onEmailGenerate', 0],
          EmailEvents::EMAIL_ON_DISPLAY => ['onEmailGenerate', 0],
        ];
      }
    
      public function onEmailBuild(EmailBuilderEvent $event): void
      {
        $event->addToken('{my_custom_token}', 'My Custom Token');
      }
    
      public function onEmailGenerate(EmailSendEvent $event): void
      {
        $event->addToken('{my_custom_token}', 'Hello <b>World!</b>');
      }
    }

.. Note::
    If you need more complex replacing you can use the event's ``$event->getContent()`` and ``$event->setContent()`` methods. See the example in the following section.

Email A/B testing

> -----------------

Reviewer 3: Heading nesting needs fixing here too, I think perhaps it should be an H2 heading?

While Mautic supports :xref:`A/B testing` out of the box, you might have more complex needs to determine A/B test winner criteria.

- ``$event->addAbTestWinnerCriteria()`` allows you to do exactly that. Using your custom logic, you can decide the winner of such criteria. An example is shown below.
- ``$event->setAbTestResults()`` is where you set the actual A/B test results. More details are in the code example below.

A/B Testing Example

> ^^^^^^^

Reviewer 3: Heading nesting needs fixing, and it must reach the end of the text for the heading.
